### PR TITLE
BZ-64644 throw an Idle Session Event on idle timeout

### PR DESCRIPTION
**Pull request by sakshamverma**
Bug ID Bug ID: https://bz.apache.org/bugzilla/show_bug.cgi?id=64644

The demo is also present there.
I found a bug while working on my product of IoT platform.
On network outage, the devices still showed connected due to the fact that tomcat didn't expire the existing connection.
KIndly review.

### CODE REVIEW

File: java/javax/websocket/OnIdleSession.java [hunk 1]
```diff
@@ -0,0 +1,12 @@
+package javax.websocket;
```
Comment (martin-g) [inline java/javax/websocket/OnIdleSession.java:1]: I'm afraid this is not part of WebSocket specification: https://www.jcp.org/en/jsr/detail?id=356 and https://download.oracle.com/otndocs/jcp/websocket-1_1-mrel-eval-spec/index.html
So you cannot add it to `javax.websocket` package

File: java/javax/websocket/Session.java [hunk 2]
```diff
@@ -77,6 +77,19 @@
...
+     *         false if the session will remain open and an IdleSession event is thrown instead
+     */
+    boolean getCloseOnIdleTimeout();
+
+    /**
+     * Get the flag to decide whether the session will be closed if idle Time is passed
+     * @param clsoeOnIdleTimeout- true for closing the session on idle timeout
+     *                            false for keeping the session open and throwing IdleSession event instead.
+     */
+    void setCloseOnIdleTimeout(boolean clsoeOnIdleTimeout);
```
Comment (martin-g) [inline java/javax/websocket/Session.java:92]: typo in the parameter name

File: java/org/apache/tomcat/websocket/WsSession.java [hunk 3]
```diff
@@ -98,6 +99,8 @@
     private volatile int maxTextMessageBufferSize = Constants.DEFAULT_BUFFER_SIZE;
     private volatile long maxIdleTimeout = 0;
     private volatile long lastActive = System.currentTimeMillis();
```
Comment (martin-g) [inline java/org/apache/tomcat/websocket/WsSession.java:101]: Probably this one should be renamed to `lastWrite` ?!

File: java/org/apache/tomcat/websocket/WsSession.java [hunk 4]
```diff
@@ -98,6 +99,8 @@
...
+    private volatile long lastActiveRead = System.currentTimeMillis();
```
Comment (martin-g) [inline java/org/apache/tomcat/websocket/WsSession.java:102]: `lastRead` ?

File: java/org/apache/tomcat/websocket/WsSession.java [hunk 5]
```diff
@@ -809,20 +825,36 @@ protected void updateLastActive() {
...
+        if (isWriteTimeout || isReadTimeout) {
+        	if(closeOnIdle) {
+                String msg = sm.getString("wsSession.timeout", getId());
+                if (log.isDebugEnabled()) {
+                    log.debug(msg);
+                }
+                doClose(new CloseReason(CloseCodes.GOING_AWAY, msg),
+                        new CloseReason(CloseCodes.CLOSED_ABNORMALLY, msg));
+        	} else {
+        		String msg = sm.getString("wsSession.timeout", getId());
```
Comment (martin-g) [inline java/org/apache/tomcat/websocket/WsSession.java:850]: indentation is off

File: java/javax/websocket/OnIdleSession.java (hunk 1 above)
Comment (sakshamverma) [reply java/javax/websocket/OnIdleSession.java:1]: Would it be okay if I move it to the package org.apache.tomcat.websocket?
And what about the other chagnes in the javax package?

**Comment by markt-asf** [comment]
There needs to be a *really* good reason to essentially introduce a Tomcat specific extension to the Jakarta WebSocket specification and thie PR hasn't (yet) provided such a justification.
If there is a good reason for adding this feature then it would be better added via the Jakarta WebSocket project where the various edge cases can be addressed.
My reading of this patch is once the idle event has been fired once, it will fire every time `WsSession.checkExpiration()` is called unless the associated handler sends a message. That seems pointless so I assume the intention is that the handler will send a message when the event fires. The end result is that a message will be sent at least every timeout milliseconds and the session will never timeout. There are simpler ways to ensure an endpoint sends a message every x milliseconds.
Absent a suitable justification for adding the feature, this PR is likely to be closed.

**Comment by sakshamverma** [comment]
@markt-asf 
I have tried to explain the problem in the video: https://youtu.be/BJf5-BpkVX8

Tomcat is not expiring the websocket channel even if there is no network. 
The way any application makes use of WS is to have some successful response from the other endpoint after sending a message.
So, if the idle session event is thrown that would mean that the network is not there. 

And the default behavior is also not changed. Its upto the products, when do they want to close the connection.

Let me know if we can have a zoom meeting to give another demo if required to understand the problem better.

**Comment by markt-asf** [comment]
The proposed approach is invasive, would require significant changes to applications for them to make use of it and - effectively - extends the Jakarta WebSocket API in a proprietary way.  I am therefore closing this PR as the approach proposed is not going to be adopted in Tomcat without it first being included in the WebSocket specification.
I have proposed a simpler approach in https://bz.apache.org/bugzilla/show_bug.cgi?id=64644 and discussion should continue there.

**Comment by sakshamverma** [comment]
@markt-asf 

I have raised another PR https://github.com/apache/tomcat/pull/346 as per your suggestion. 

Could you please let me know what are the next steps for it to merge? It will be a great help for our product (Thingworx) which uses tomcat as the webserver.

**Pull request closed without merging**